pennylane/interfaces/batch/jax.py

```diff
@@ -1,299 +1,332 @@
+
+        for t, a in zip(tapes, params):
+            new_tapes.append(t.copy(copy_operations=True))
+            new_tapes[-1].set_parameters(a)
+
+        with qml.tape.Unwrap(*new_tapes):
+            res, jacs = execute_fn(new_tapes, **gradient_kwargs)
+
+        if len(res) > 1:
+            res, jacs = [jnp.array(r) for r in res], [jnp.array(j) for j in jacs]
+        else:
+            res, jacs = jnp.array(res), jnp.array(jacs)
+        return res, jacs
+
+    def wrapped_exec_fwd(params):
+        res, jacs = wrapped_exec(params)
+        return res, tuple([jacs, params])
+
+    def wrapped_exec_bwd(params, g):
+
+        # Use the jacobian that was computed on the forward pass
+        jacs, params = params
+
+        # Note: there is an issue when jax.jacobian is being used, either due
+        # to issues with tensor updating (TypeError: Updates tensor must be of
+        # rank 0; got 1) or because jax.vmap introduces a redundant
+        # dimensionality in the result by duplicating entries
+        # Example to the latter:
+        #
+        # 1. Output when using jax.jacobian:
+        # DeviceArray([[-0.09983342,  0.01983384],\n
+        #              [-0.09983342, 0.01983384]], dtype=float64),
+        # DeviceArray([[ 0.        , -0.97517033],\n
+        #              [ 0.        , -0.97517033]], dtype=float64)),
+        #
+        # 2. Expected output:
+        # DeviceArray([[-0.09983342, 0.01983384],\n
+        #             [ 0.        , -0.97517033]]
+        #
+        # The output produced by this function matches 2.
+        scalar_outputs = all(t.output_dim == 1 for t in tapes)
+        if not scalar_outputs:
+            raise ValueError(
+                "Computing the jacobian of vector-valued tapes is not supported currently in forward mode."
+            )
+
+        # Adjust the structure of how the jacobian is returned to match the
+        # non-forward mode cases
+        # E.g.,
+        # [DeviceArray([[ 0.06695931,  0.01383095, -0.46500877]], dtype=float32)]
+        # is mapped to
+        # [[DeviceArray(0.06695931, dtype=float32), DeviceArray(0.01383095,
+        # dtype=float32), DeviceArray(-0.46500877, dtype=float32)]]
+        res_jacs = []
+        for j in jacs:
+            this_j = []
+            for i in range(j.shape[1]):
+                arr = (
+                    j[0, i] if j.shape[0] == 1 else jnp.array([j[k, i] for k in range(j.shape[0])])
+                )
+                this_j.append(arr)
+            res_jacs.append(this_j)
+        return tuple([tuple(res_jacs)])
+
+    wrapped_exec.defvjp(wrapped_exec_fwd, wrapped_exec_bwd)
+    res = wrapped_exec(params)
+
+    tracing = any(isinstance(r, jax.interpreters.ad.JVPTracer) for r in res)
+
+    # When there are no tracers (not differentiating), we have the result of
+    # the forward pass and the jacobian, but only need the result of the
+    # forward pass
+    if len(res) == 2 and not tracing:
+        res = res[0]
```

Comment:
@antalszava could this be the reason for the duplication during the backwards pass? I would think that this function should return `res[0]` always?

Reply:
Returning it always will raise errors like `IndexError: Too many indices for array: 1 non-None/Ellipsis indices for dim 0`.

Oddly enough, when moving the output check
```python
        scalar_outputs = all(t.output_dim == 1 for t in tapes)
```
to have
```python
    if len(res) == 2 and not tracing or not scalar_outputs:
```
the `test_jacobian` and other QNode examples pass! However, the simple tape example `test_multiple_expvals_raises_fwd_device_grad` fails with:
`TypeError: Updates tensor must be of rank 0; got 1.`

For now, I'd rather disallow this case until we've investigated it more.